src/realm/sync/noinst/client_impl_base.hpp

```diff
@@ -56,6 +56,57 @@ class SessionWrapperStack {
     SessionWrapper* m_back = nullptr;
 };
 
+/// Register a function whose invocation can be triggered repeatedly.
+///
+/// While the function is always executed by the event loop thread, the
+/// triggering of its execution can be done by any thread.
+///
+/// The function is guaranteed to not be called after the Trigger object is
+/// destroyed.
+///
+/// Note that even though the trigger() function is thread-safe, the Trigger
+/// object, as a whole, is not. In particular, construction and destruction must
+/// not be considered thread-safe.
+class Trigger : public std::enable_shared_from_this<Trigger> {
```

Comment:
Let's move Trigger into the _socket_provider.hpp_ file since it's related to that work and so _server.cpp_ doesn't need to include client_impl_base.
Maybe add a `SyncSocketProvider::create_trigger()` function that passes a weak reference to the SyncSocketProvider for calling `post()`.
We can then make the Trigger class final since we don't expect any implementations to replace this functionality.

Reply:
Let's move this into its own file and not into the socket_provider.hpp file and include that in the places we need to include it.

Reply:
Also, if you make this inherit from util::AtomicRefCountBase<> you can use it with bind_ptr's, which are a little more lightweight.